src/http-resolver.js

```diff
@@ -0,0 +1,66 @@
+const http2 = require('http2-wrapper');
+
+class HttpResolver {
+    constructor() {
+        this._cache = new Map();
+        this._maxCacheSize = 1000;
+    }
+
+    /**
+     *
+     * @param {URL} parsedUrl
+     * @param {boolean} rejectUnauthorized
+     * @returns {string} resolved protocol
+     */
+    async resolveHttpVersion(parsedUrl, rejectUnauthorized) {
+        const { hostname, port } = parsedUrl;
+        const cacheKey = `${hostname}:${port}`;
+
+        let httpVersion = this._getFromCache(cacheKey);
+
+        if (!httpVersion) {
+            const result = await http2.auto.resolveProtocol({
+                host: hostname,
+                servername: hostname,
+                port,
+                ALPNProtocols: ['h2', 'http/1.1'],
+                rejectUnauthorized,
+            });
+            httpVersion = result.alpnProtocol;
+            this._maybeRemoveOldestKey();
```

Comment:
this could be called in the `setToCache` function, no? The `resolveHttpVersion` function does not care about cache management